Disable webkit separately from gtk integration

mid-kid opened the pull request.

ilya-fedin commented:
I don't think that's really needed

ilya-fedin commented on variables.cmake:
```diff
@@ -32,6 +32,7 @@ option(DESKTOP_APP_DISABLE_DBUS_INTEGRATION "Disable all code for D-Bus integrat
 option(DESKTOP_APP_DISABLE_X11_INTEGRATION "Disable all code for X11 integration (Linux only)." OFF)
 option(DESKTOP_APP_DISABLE_WAYLAND_INTEGRATION "Disable all code for Wayland integration (Linux only)." OFF)
 option(DESKTOP_APP_DISABLE_GTK_INTEGRATION "Disable all code for GTK integration (Linux only)." OFF)
+option(DESKTOP_APP_DISABLE_WEBKIT_INTEGRATION "Disable all code for WebkitGTK integration (Linux only)." OFF)
```
GTK integration is integration because it integrates to gtk-based DEs. WebKit doesn't integrate to anything.

mid-kid replied:
I see, I'll change that.

ilya-fedin commented on options.cmake:
```diff
@@ -49,6 +49,13 @@ if (DESKTOP_APP_DISABLE_GTK_INTEGRATION)
     )
 endif()
 
+if (DESKTOP_APP_DISABLE_WEBKIT_INTEGRATION)
+    target_compile_definitions(common_options
+    INTERFACE
+        DESKTOP_APP_DISABLE_WEBKIT_INTEGRATION
+    )
+endif()
+
```
well...

ilya-fedin commented:
(the fact these sugesstions applied doesn't mean it became needed)

mid-kid commented:
True, that's up to you to decide. I'll be applying this patch to gentoo to spare some people the trouble of building webkitgtk, since it needs about 5x the build time of tdesktop for a feature I can't imagine many people will use, while keeping the gtk integration (font/icon settings right now, afaict). (note that people can always enable it if they so choose). I'll be supporting it as long as it remains feasible (simple patch and all), and this PR is simply an attempt to get rid of the downstream patch. It might be unnecessary upstream and for many distributions, but for gentoo I consider it necessary, and maybe for other distros that either can't support webkit (architecture/platform/whatever) or can't spare the 100MB webkitgtk install size.

While I understand trying to keep the build matrix and compilation options few (given a CI run already needs multiple hours to finish), it doesn't make much sense to me to group gtk integration (desktop support) along with webkit support (app feature), even if the latter needs gtk, too.

ilya-fedin commented:
> font/icon settings right now, afaict

All that work does Qt's integration. tdesktop has this duplication for static binary that should use dlopen for loading gtk library.

ilya-fedin commented:
And tdesktop doesn't use system font settings at all, open sans is forced.

mid-kid commented:
Oh, I see. So this GTK integration is unnecessary except for static builds? What about [image copy-paste](https://gitweb.gentoo.org/repo/gentoo.git/tree/net-im/telegram-desktop/telegram-desktop-2.7.1-r2.ebuild#n136)? I'll test that with USE=-gtk when it finishes building, but I was under the impression that GTK integration was useful to have enabled.

ilya-fedin commented:
> So this GTK integration is unnecessary except for static builds?

AFAIK the only part where there's no other backend for asking a setting is decoration layout. It asks only from gtk and fallbacks to minimize, maximize, close on right side.

> What about [image copy-paste](https://gitweb.gentoo.org/repo/gentoo.git/tree/net-im/telegram-desktop/telegram-desktop-2.7.1-r2.ebuild#n136)?

Sometimes I regret that I added this hack. People stopped to complain to the Qt bugreport and Qt folks stopped looking into it. Looks like there's no other such popular Qt application that has clipboard interaction as a primary feature.

mid-kid commented:
Alright, I find that enough reason to allow people to enable GTK integration even without webkit, at least as long as these issues exist.

Weird though that a UI toolkit this big and popular doesn't support something as universal as image copy/paste... Would be nice if that were to be fixed in general, as I know some other applications that could use it.

ilya-fedin commented:
> Weird though that a UI toolkit this big and popular doesn't support something as universal as image copy/paste...

Why do you say it doesn't support image copy/paste? It support image copy/paste, but there are a bug in xcb backend that clipboard data reading stops earlier than entire data is read, so Qt can't paste big images on X11, leading to crash of application where you copied that image from if it doesn't handle X exceptions (e.g. all gtk2 applications, including gimp).

mid-kid commented:
I see, sorry, I misunderstood your message. Yeah, that seems like a very annoying bug, though thankfully gtk2 applications are few and far between these days, with GIMP being practically the only gtk2 application I have left on my system (and even that's finally being ported).

Still though, if this is the solution for now, I think it's worth keeping the gtk integration, at least until it's solved on the qt side or a patch appears that can be applied system-wide (and distributions are fine with including it). Do you happen to have a link to the Qt bug report about this?

ilya-fedin commented:
> Do you happen to have a link to the Qt bug report about this so I can track it?

The link is already in your ebuild. There are also an issue that feels like a duplicate: https://bugreports.qt.io/browse/QTBUG-79656 (image is there, so hasImage is true, but on actual paste attempt the data is not fully received and imageData is null)

mid-kid commented:
Note (sorry for the bump): I just found that building with `-DDESKTOP_APP_DISABLE_WEBKIT=OFF -DDESKTOP_APP_DISABLE_GTK_INTEGRATION=ON` fails due to LoadLibrary() not being available. While it's fixable with an [additional patch](https://pastebin.com/raw/YpYpSF2b), running such a configuration makes no sense to me at all, given webkitgtk's inherent dependence on gtk. Maybe this can be enforced by the cmakefiles...

ilya-fedin commented:
Yeah, DESKTOP_APP_DISABLE_WEBKIT=OFF (I think DESKTOP_APP_DISABLE_WEBKIT2GTK is better though since e.g. macOS has CocoaWebKit) should throw a fatal error when DESKTOP_APP_DISABLE_GTK_INTEGRATION is ON, but I didn't say about that since it's unlikely to be merged anyway

mid-kid commented:
This patch required updating for v2.8.8+, but given the discussion above, I won't be updating this PR (or the linked one) until a decision is made.

Newer versions of the patch can be found here: https://github.com/gentoo/gentoo/tree/master/net-im/telegram-desktop/files (Look for "disable-webkit-separately"). Note that the patch as hosted here isn't a direct proposal for inclusion into upstream, as the patch is kept minimal to make downstream maintainance easier. In the event that the addition of this build flag is accepted, we can try to work out a proper patch.

ilya-fedin commented:
> This patch required updating for v2.8.8+, but given the discussion above, I won't be updating this PR (or the linked one) until a decision is made.

Well, personally I'm starting to think that this option is needed looking at how long webkitgtk builds on flatpak infrastructure. That's probably a nightmare on desktops and notebooks. But this needs a permission from @john-preston.

ilya-fedin commented:
Also, there's some progress on the clipboard problem (finally, after 5 years!), I tested the patch and it works perfectly for me, it's already in tdesktop's Qt patchset.
https://codereview.qt-project.org/c/qt/qtbase/+/306771

As soon as they will merge it, I think it would be a time to remove the hack from tdesktop, then everyone who experience pasting problems can apply the patch themselves.

ilya-fedin commented:
Oh, I see you disabled jemalloc, that's a bad decision. glibc's allocator works really badly with tdesktop and creates feeling that tdesktop leaks and never releases memory. I don't really recommend do that as that will make a poor UX for your package users.

mid-kid commented:
I see! Thanks for commenting on that. I'll adapt the patch message and include a post-install warning about it. Though I wonder why this is the case? Is this an issue with different apps or just tdesktop?

ilya-fedin commented:
>  Is this an issue with different apps or just tdesktop?

What different apps are you talking about? All I know is that there always were complains about RAM usage from Linux users. @john-preston reworked media RAM storage multiple time, but no lack. When I found mention about `@tdesktop_crash` somewhere (looks like the one who created it wanted to report about the 32-bit RAM limit on Windows, but since I'm on Linux...), I reproduced that reliably for first time. When I scroll the to the top and close the channel with ESC, the memory is not released. Then I asked @john-preston to repeat that, he did that on Windows and... the memory was released. Then, I tried to experiment with other allocators. Initially jemalloc worked strangely and I bundled musl's mallocng (from the draft repo). Once jemalloc devs explained that it's due to the absence of background thread, I switched tdesktop to jemalloc since mallocng is maintained only as a part of musl now and I don't really want to monitor changes for backporting.

mid-kid commented:
Thanks for explaining. With "different apps" I meant anything else that uses malloc. The way you describe the issue sounds like it'd be a glibc bug, so I was wondering if anything else would be affected. Surely enough, I've tried what you said, and it's true. Interestingly, however, the memory use increases as you scroll through more chats, but caps out at just under 1GB on my machine (24GB installed RAM), and doesn't allocate more no matter how many more chats are scrolled through. I wonder if glibc has some sort of cache...

As for the Qt patch, unfortunately I don't think that's going to land in any public release, as Qt won't be providing patched versions of 5.15 for free anymore. KDE is maintaining fixes for it, but I haven't looked into the infrastructure of it yet (to know how to submit one), nor do I know if they're going to provide release tarballs. It might take a while before it lands in most distros, unless the patch is adopted through other means.

ilya-fedin commented:
> but caps out at just under 1GB on my machine (24GB installed RAM)

There are reports about 2GB, 4GB RAM usages...

> It might take a while before it lands in most distros, unless the patch is adopted through other means.

I would say I don't really care about that. If people will report about clibpoard problems, I'll suggest them to install one of official builds or to go to distro bugtracker and ask to backport the patch. That hack is really ugly and I really want to get rid of it as soon as possible.

ilya-fedin commented:
> There were reports about 2GB, 4GB RAM usages...

e.g.: https://github.com/telegramdesktop/tdesktop/issues/16254

ilya-fedin commented:
I guess it depends on actual content, I don't thinnk these owls are biggest content at Telegram

mid-kid commented:
Nevermind, it got much higher for me, too. Thanks for the issue link.

ilya-fedin commented:
> Thanks for the issue link.

It's really simple to find a lot of such issues ;)
https://github.com/telegramdesktop/tdesktop/issues?q=is%3Aissue+is%3Aopen+ram

ilya-fedin commented:
@mid-kid can you rebase, rename the option to DESKTOP_APP_DISABLE_WEBKITGTK and add a check (with cmake's fatal_error or with `#error` in sources) that GTK is not disabled when WEBKITGTK enabled?

mid-kid commented:
@ilya-fedin I've rebased and fixed this PR as well as the attached lib_webview PR. However, [that's not all that's required for this patchset](https://github.com/gentoo/gentoo/blob/master/net-im/telegram-desktop/files/tdesktop-2.8.9-disable-webkit-separately.patch#L14). I'm just unsure how to tackle the remaining parts in a way that's satisfying for telegram upstream.

ilya-fedin commented:
> However, [that's not all that's required for this patchset](https://github.com/gentoo/gentoo/blob/master/net-im/telegram-desktop/files/tdesktop-2.8.9-disable-webkit-separately.patch#L15-L50).

I see only one error:

```
+#ifndef DESKTOP_APP_DISABLE_WEBKIT
 	} else if (type == Type::Webview) {
 		Webview::WebKit2Gtk::SetServiceName(
 			kWebviewService.utf16().arg(h).arg("%1").toStdString());
 
 		return;
+#endif
```
->
```
 	} else if (type == Type::Webview) {
+#ifndef DESKTOP_APP_DISABLE_WEBKITGTK
 		Webview::WebKit2Gtk::SetServiceName(
 			kWebviewService.utf16().arg(h).arg("%1").toStdString());
+#endif
 
 		return;
```

Other than that, it's good (maybe you want to ifndef webkit include in the file as well, though)

mid-kid commented:
I was honestly expecting I'd have to add functions to [webview_linux.cpp](https://github.com/desktop-app/lib_webview/blob/master/webview/platform/linux/webview_linux.cpp) to avoid the ifdefs elsewhere, but alright. I'll do this later real quick. Should `GtkIntegration::Start` return early when type = webview and it's disabled, or can I rely on the return after SetServiceName?

ilya-fedin commented:
> I was honestly expecting I'd have to add functions to [webview_linux.cpp](https://github.com/desktop-app/lib_webview/blob/master/webview/platform/linux/webview_linux.cpp) to avoid the ifdefs elsewhere, but alright.

you won't be able to get rid of them since that would require calling these methods from Webview namespace that should be... commented, again.

ilya-fedin commented on variables.cmake:
```diff
@@ -37,6 +37,7 @@ option(DESKTOP_APP_USE_PACKAGED "Find libraries using CMake instead of exact pat
 option(DESKTOP_APP_USE_PACKAGED_LAZY "Bundle recommended Qt plugins for self-contained packages. (Linux only)" OFF)
 option(DESKTOP_APP_USE_PACKAGED_FFMPEG_STATIC "Link ffmpeg statically in packaged mode." OFF)
 option(DESKTOP_APP_DISABLE_SPELLCHECK "Disable spellcheck library." ${osx_special_target})
+option(DESKTOP_APP_DISABLE_WEBKITGTK "Disable WebkitGTK library (Linux only)." OFF)
```
```suggestion
option(DESKTOP_APP_DISABLE_WEBKITGTK "Disable WebKitGTK library (Linux only)." OFF)
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/desktop-app/cmake_helpers?pullRequest=91) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/desktop-app/cmake_helpers?pullRequest=91) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: mid-kid<br/>:x: john-preston<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/desktop-app/cmake_helpers?pullRequest=91) it.</sub>
